src/hotspot/share/prims/jvm.cpp

```diff
@@ -3711,6 +3727,69 @@ JVM_ENTRY(void, JVM_LogLambdaFormInvoker(JNIEnv *env, jstring line))
 #endif // INCLUDE_CDS
 JVM_END
 
+#if INCLUDE_CDS
+class DumpClassListCLDClosure : public CLDClosure {
+  fileStream *_stream;
+public:
+  DumpClassListCLDClosure(fileStream* f) : CLDClosure() { _stream = f; }
+  void do_cld(ClassLoaderData* cld) {
+    for (Klass* klass = cld->klasses(); klass != NULL; klass = klass->next_link()) {
+      if (klass->is_instance_klass()) {
+        InstanceKlass* ik = InstanceKlass::cast(klass);
+        if (ik->is_shareable()) {
+          _stream->print_cr("%s", ik->name()->as_C_string());
+        }
+      }
+    }
+  }
+};
+#endif
+
+JVM_ENTRY(void, JVM_DumpClassListToFile(JNIEnv *env, jstring listFileName))
+#if INCLUDE_CDS
+  ResourceMark rm(THREAD);
+  Handle file_handle(THREAD, JNIHandles::resolve_non_null(listFileName));
+  char* file_name  = java_lang_String::as_utf8_string(file_handle());
+  fileStream stream(file_name, "w");
+  if (stream.is_open()) {
+    MutexLocker lock(ClassLoaderDataGraph_lock);
+    DumpClassListCLDClosure collect_classes(&stream);
+    ClassLoaderDataGraph::loaded_cld_do(&collect_classes);
+  } else {
+    THROW_MSG(vmSymbols::java_io_IOException(), "Failed to open file");
+  }
+#endif // INCLUDE_CDS
+JVM_END
+
+JVM_ENTRY(void, JVM_DumpDynamicArchive(JNIEnv *env, jstring archiveName))
+#if INCLUDE_CDS
+  assert(UseSharedSpaces && RecordDynamicDumpInfo, "Sanity check");
```

Comment:
I think the message should say why this is true.

How about `assert(UseSharedSpaces && RecordDynamicDumpInfo, "already checked in arguments.cpp");`?

(same for the next assert statement at line 3774)